### PR TITLE
Added delivery options to broker create

### LINKED ISSUE #1557: Add delivery options for "kn broker create"
### Feature request

Actually, the support for broker options is quite limited, only the broker class can be currently specified, but none of the delivery options that are described in https://github.com/knative/eventing/blob/5e3183b552cabe35ae1015fa1baff5cd67c0b7cf/pkg/apis/duck/v1/delivery_types.go#L31-L84

These options are independent of the broker implementation and can be typed (.e.g `--dl-sink`). This is different from https://github.com/knative/client/issues/1525 where a reference to an implementation-specific configuration should be able to add.

### Use case

Creating a broker with delivery options that are different from the default values.

### UI Example

The following options are proposals, the detailed naming is up to discussion:

```
--dl-sink <sink reference>              Reference to a sink for delivering events that can not be sent
--retry <retries>.                      Number of retries before sending the event to a dead-letter sink
--timeout <duration>                    Timeout for a single request
--backoff-policy  <policy>              Backoff policy for retries, either "linear" or "exponential"
--backoff-delay <duration>              Based delay between retries
--retry-after-max <duration>            Upper bound for a duration specified in an "Retry-After" header (experimental)
```

### PR DESCRIPTION
**Pull request by vyasgun**
## Description
Added delivery options to `kn broker create`

## Changes

* Added `--dl-sink`, `--timeout`, --`backoff-policy`, `backoff-delay`, `retry-after-max`, `--retry` flags
* Validation is on the server side for the values
* Added unit test cases

## Reference

Fixes #1557

### CODE REVIEW

File: pkg/eventing/v1/client_test.go [hunk 1]
```diff
@@ -518,6 +524,23 @@ func newBrokerWithClass(name string) *eventingv1.Broker {
...
+	testTimeout := "PT10S"
+	retry := int32(2)
+	policy := v1.BackoffPolicyType("linear")
+	return NewBrokerBuilder(name).
+		Namespace(testNamespace).
+		DlSink(sink).
+		Timeout(&testTimeout).
+		Retry(&retry).
+		BackoffDelay(&testTimeout).
+		BackoffPolicy(&policy).
```
Comment (dsimansk) [inline pkg/eventing/v1/client_test.go:540]: Missing `RetryAfterMax()` call.

File: pkg/eventing/v1/client.go [hunk 2]
```diff
@@ -387,6 +386,80 @@ func (b *BrokerBuilder) Class(class string) *BrokerBuilder {
 	return b
 }
 
+// DlSink for the broker builder
+func (b *BrokerBuilder) DlSink(dlSink *duckv1.Destination) *BrokerBuilder {
+	empty := duckv1.Destination{}
+	if dlSink == nil || *dlSink == empty {
+		return b
```
Comment (dsimansk) [inline pkg/eventing/v1/client.go:393]: For the sake of managing test coverage, all those corner cases would ideally be covered. IMO `nil` cases should be fine for us.

File: pkg/kn/commands/broker/delivery_option_flags.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	RetryCount    int32
+	Timeout       string
+	BackoffPolicy string
+	BackoffDelay  string
+	RetryAfterMax string
+}
+
+func (d *DeliveryOptionFlags) Add(cmd *cobra.Command) {
+	d.SinkFlags.AddWithFlagName(cmd, "dl-sink", "")
+	cmd.Flag("dl-sink").Usage = "Reference to a sink for delivering events that can not be sent"
```
Comment (dsimansk) [inline pkg/kn/commands/broker/delivery_option_flags.go:33]: Well I'm not sure if shouldn't keep the default usage with all the explanation of sinks URIs.

File: pkg/kn/commands/broker/delivery_option_flags.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+	cmd.Flags().Int32Var(&d.RetryCount, "retry", 0, "Number of retries before sending the event to a dead-letter sink")
```
Comment (dsimansk) [inline pkg/kn/commands/broker/delivery_option_flags.go:35]: Is there any default enforced on the Eventing side, that we could potentially follow here?

**Review by dsimansk — CHANGES_REQUESTED** [review]
Added a few comments around coverage and help messages, but otherwise looks OK.

File: pkg/kn/commands/broker/delivery_option_flags.go (hunk 4 above)
Comment (dsimansk) [reply pkg/kn/commands/broker/delivery_option_flags.go:35]: To be clear, I meant for all the flags here, not just the `retry` one.

File: pkg/eventing/v1/client.go (hunk 2 above)
Comment (dsimansk) [reply pkg/eventing/v1/client.go:393]: > IMO nil cases should be fine for us.

I used a very confusing construct in my first comment, sorry about that :)

It should really go like this. Let's add test cases with `nil` values to cover return of builder without any changes. It's for the sake of coverage mainly, but it is what it is. :)

File: pkg/eventing/v1/client_test.go (hunk 1 above)
Comment (vyasgun) [reply pkg/eventing/v1/client_test.go:587]: Thanks, fixed.

File: pkg/kn/commands/broker/delivery_option_flags.go (hunk 4 above)
Comment (vyasgun) [reply pkg/kn/commands/broker/delivery_option_flags.go]: Thanks, I updated the usages.

File: pkg/eventing/v1/client.go (hunk 2 above)
Comment (vyasgun) [reply pkg/eventing/v1/client.go:393]: I added the `nil` test cases.

File: pkg/kn/commands/broker/delivery_option_flags.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	cmd.Flags().Int32Var(&d.RetryCount, "retry", 0, "The minimum number of retries the sender should attempt when "+
+		"sending an event before moving it to the dead letter sink.")
+	cmd.Flags().StringVar(&d.Timeout, "timeout", "", "The timeout of each single request. The value must be greater than 0.")
+	cmd.Flags().StringVar(&d.BackoffPolicy, "backoff-policy", "", "The retry backoff policy (linear, exponential).")
+	cmd.Flags().StringVar(&d.BackoffDelay, "backoff-delay", "", "The delay before retrying.")
+	cmd.Flags().StringVar(&d.RetryAfterMax, "retry-after-max", "", "An optional upper bound on the duration specified in a "+
+		"\"Retry-After\" header when calculating backoff times for retrying 429 and 503 response codes. "+
+		"Setting the value to zero (\"PT0S\") can be used to opt-out of respecting \"Retry-After\" header values altogether. "+
+		"This value only takes effect if \"Retry\" is configured, and also depends on specific implementations (Channels, Sources, etc.) "+
+		"choosing to provide this capability.")
```
Comment (dsimansk) [inline pkg/kn/commands/broker/delivery_option_flags.go:44]: Wrt to `--retry` + `--retry-after-maxt` is there validation provided on the webhook side?

**Review by dsimansk — COMMENTED** [review]
Added a short comment about validation.

There's another angle of this change I haven't noticed at first. Originally broker command groups doesn't include `update`. There wasn't no options to modify, except from create/delete flow. However, it should be added probably with this addition flags.

File: pkg/kn/commands/broker/delivery_option_flags.go (hunk 5 above)
Comment (vyasgun) [reply pkg/kn/commands/broker/delivery_option_flags.go:44]: Yes, the validation is present.

**Comment by vyasgun** [comment]
@dsimansk Sorry about the delay. I just added the update command to this PR.
